File: telegraf_resource_monitor/telegraf_resource_monitor/sensor_message_publisher.py
from rclpy._rclpy_pybind11 import RCLError
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data

from resource_monitoring_interfaces.msg import Field, Resource
from telegraf_resource_monitor.sensor_message import SensorMessage


class SensorMessagePublisher:
    def __init__(self, node: Node, message: SensorMessage) -> None:
        self.node = node
        self.logger = node.get_logger()

        sensor_type = message.name
        sensor_tags = message.tags

        # Convert tags dict to a pathed string
        tags_str = (
            "/".join(f"{value}" for value in sensor_tags.values())
            if sensor_tags
            else ""
        )

        topic_name = f"{sensor_type}/{tags_str}"

        # sanitize topic name to avoid issues with special characters
        topic_name = topic_name.replace(" ", "_").replace("-", "_")

        # remove trailing slashes
        topic_name = topic_name.rstrip("/")
        topic_name = topic_name.lower()

        self.logger.info(f"creating publisher for resource {topic_name}")

        self.publisher = node.create_publisher(
            msg_type=Resource,
            topic=topic_name,
            qos_profile=qos_profile_sensor_data,
        )

    def publish(self, message: SensorMessage) -> None:
        fields = []

        for field_name, field_value in message.fields.items():
            field_msg = Field()
            field_msg.name = field_name
            field_msg.value = float(field_value)
            fields.append(field_msg)

        resource_msg = Resource()
        resource_msg.header.stamp = self.node.get_clock().now().to_msg()
        resource_msg.fields = fields

        try:
            self.publisher.publish(resource_msg)
        except RCLError as e:
            if "publisher's context is invalid" in str(e):
                return

File: telegraf_resource_monitor_cpp/src/sensor_message_publisher.cpp


#include <cctype>
#include <map>
#include <memory>
#include <string>
#include <string_view>
#include <utility>

#include "rclcpp/node.hpp"
#include <builtin_interfaces/msg/time.hpp>

#include <fmt/format.h>

#include "resource_monitoring_interfaces/msg/field.hpp"
#include "resource_monitoring_interfaces/msg/resource.hpp"
#include "telegraf_resource_monitor_cpp/sensor_message.hpp"
#include "telegraf_resource_monitor_cpp/sensor_message_publisher.hpp"

using Field = resource_monitoring_interfaces::msg::Field;
using Time = builtin_interfaces::msg::Time;
using TagsKey = std::map<std::string, std::string>;

SensorMessagePublisher::SensorMessagePublisher(const rclcpp::Node::SharedPtr& node,
                                               std::string_view sensor_type,
                                               const TagsKey& tag_keys)
  : node_{node}
  , logger_{node->get_logger()}
{
  std::string topic_name{create_topic_name(sensor_type, tag_keys)};

  publisher_ptr_ = node->create_publisher<ResourceType>(topic_name, 10);

  logger_.info("made resource publisher for {}", topic_name);
}

std::string SensorMessagePublisher::create_topic_name(std::string_view sensor_type,
                                                      const TagsKey& sensor_tags)
{
  if (sensor_tags.empty())
  {
    return std::string(sensor_type);
  }

  std::string topic_str{combine_type_and_tags(sensor_type, sensor_tags)};
  sanitize_topic_name(topic_str);
  return topic_str;
}

std::string SensorMessagePublisher::combine_type_and_tags(std::string_view sensor_type,
                                                          const TagsKey& sensor_tags)
{
  std::string topic_str{};

  bool first = true;
  for (const auto& [key, val] : sensor_tags)
  {
    if (!first)
    {
      topic_str.append("/");
    }
    topic_str.append(val);
    first = false;
  }
  return std::string(sensor_type) + "/" + topic_str;
}

void SensorMessagePublisher::sanitize_topic_name(std::string& topic_str)
{
  for (char& c : topic_str)
  {
    if (!std::isalnum(c) && c != '_' && c != '~' && c != '/')
    {
      c = '_';
    }
  }
}

void SensorMessagePublisher::publish(SensorMessage& message) const
{
  Time current_time{};
  current_time.set__sec(message.timestamp);

  // The message is built in a unique_ptr so it can be published by moving it. With
  // intra process communication enabled, publishing a unique_ptr hands ownership to the
  // intra process manager and subscribers in the same process read this exact instance.
  // refer to src/docs/intra_process_communication.md for more.
  auto resource = std::make_unique<ResourceType>();
  resource->header.stamp = current_time;

  for (const auto& [field_name, field_value] : message.fields)
  {
    Field field_msg{};
    field_msg.name = field_name;
    field_msg.value = field_value;

    resource->fields.emplace_back(std::move(field_msg));
  }

  // logging the address here and in the subscriber shows whether both ends touch the same
  // instance, which is the proof that intra process communication is actually being used
  logger_.debug("publishing message on {} at address {}", publisher_ptr_->get_topic_name(),
                fmt::ptr(resource.get()));

  publisher_ptr_->publish(std::move(resource));
}
